Move math.hpp from API to tt-stl

riverwuTT opened the pull request:
### Problem description
math.hpp (provides div_up, round_up, round_down) does not belong to metal's API, it is more suitable to tt-stl.
Also, there's at least three implementations of `div_up` across metal and ttnn.

### What's changed
This PR:
1. moves math.hpp to tt_stl
2. Updates associated function namespace to `ttsl::math`
3. Update all references to new namespace & header
4. Cleans up repetitive implementations and redirect them to `ttsl::math`
5. Goes through all files that uses `math.hpp` and make sure the include is appropriate.

### Downstream
TTNN: All references updated
Forge: header not used
MLIR: header not used

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/20147990210)

riverwuTT commented on tt_stl/tt_stl/math.hpp:
This is the moved file.

riverwuTT commented on tt_stl/tt_stl/math.hpp:
```diff
@@ -8,7 +8,7 @@
 #include <cstdint>
```
replace `<cassert>` with `<tt_stl/assert.hpp>` after #28426

riverwuTT commented on ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:
```diff
@@ -9,10 +9,11 @@
 
 #include <tt-metalium/buffer_types.hpp>
+#include <tt_stl/math.hpp>
 
 #include <tt_stl/assert.hpp>
 #include <tt-logger/tt-logger.hpp>
-#include "tt-metalium/math.hpp"
+#include <tt_stl/math.hpp>
```
seems to have two includes?

riverwuTT reviewed (commented):
I have manually checked all files for round 1, I've found some files that have duplicative includes, but everything else looks fine.

riverwuTT reviewed (commented):
Pass 2 looks good

riverwuTT commented on ttnn/cpp/ttnn/operations/experimental/slice_write/device/slice_write_program_factory.cpp:
```diff
@@ -10,9 +10,10 @@
 
 #include "slice_write_op.hpp"
-#include "tt-metalium/math.hpp"
+#include <tt_stl/math.hpp>
```
Good catch

blozano-tt reviewed (approved):
rm -rf math

fvranicTT commented on tt_stl/tt_stl/math.hpp:
```diff
@@ -4,11 +4,10 @@
 #include <type_traits>
 
-namespace tt {
+namespace ttsl::math {
```
I know it might be late, as you have plenty of approvals, but I think we should not have yet another top level namespace besides namespace tt. Schemes like `namespace tt::sl::math` are quite common in the industry. Check boost or Poco for example. A collection of different independent libraries are all under one top level namespace, boost (or Poco).
A single root namespace is better because it organizes, avoids collisions, and gives identity. Multiple top-level namespaces fragment the library and risk conflicts with other vendors.
Sorry I got to this only now.

blozano-tt replied:
History: it used to be tt_stl - Tenstorrent standard template library. It got shortened.

riverwuTT committed:
refactor: Move math.hpp from tt-metalium to tt_stl

- Relocate math.hpp from tt_metal/api/tt-metalium/ to tt_stl/tt_stl/
- Update namespace from tt:: to ttsl::math::
- Replace cassert with tt_stl/assert.hpp (TT_ASSERT)
- Add math.hpp to tt_stl CMakeLists.txt exports

This commit creates the foundation; consumers will be updated next.

riverwuTT committed:
refactor: Update all consumers to use ttsl::math namespace

- Update include paths from <tt-metalium/math.hpp> to <tt_stl/math.hpp>
- Change all tt::div_up/round_up/round_down to ttsl::math::*
- Fix unqualified calls to use ttsl::math:: prefix
- Remove math.hpp from tt_metal CMakeLists.txt public API list
- No behavioral changes, pure mechanical refactoring

riverwuTT commented:
The PR was abandoned, will remake one in the future.
